ffi/derive/src/bindgen.rs

```diff
@@ -0,0 +1,456 @@
+use proc_macro2::Span;
+use proc_macro_error::{abort, OptionExt};
+use quote::quote;
+use syn::{parse_quote, Ident, Type};
+
+use crate::{
+    get_ident, snake_case_ident,
+    visitor::{FnArgDescriptor, FnDescriptor},
+};
+
+pub fn gen_ffi_fn(fn_descriptor: &FnDescriptor) -> proc_macro2::TokenStream {
+    let ffi_fn_name = gen_ffi_fn_name(fn_descriptor);
+
+    let self_arg = fn_descriptor
+        .receiver
+        .as_ref()
+        .map(gen_ffi_fn_arg)
+        .map_or_else(Vec::new, |self_arg| vec![self_arg]);
+    let fn_args: Vec<_> = fn_descriptor
+        .input_args
+        .iter()
+        .map(gen_ffi_fn_arg)
+        .collect();
+    let output_arg = ffi_output_arg(fn_descriptor).map(gen_ffi_fn_arg);
+    let fn_body = gen_fn_body(fn_descriptor);
+
+    let ffi_fn_doc = format!(
+        " FFI function equivalent of [`{}::{}`]",
+        fn_descriptor.self_ty.get_ident().expect_or_abort("Defined"),
+        fn_descriptor.method_name
+    );
+
+    quote! {
+        #[doc = #ffi_fn_doc]
+        #[no_mangle]
+        pub unsafe extern "C" fn #ffi_fn_name(#(#self_arg,)* #(#fn_args,)* #output_arg) -> iroha_ffi::FfiResult {
+            #fn_body
+        }
+    }
+}
+
+fn gen_ffi_fn_name(fn_descriptor: &FnDescriptor) -> Ident {
+    let self_ty_name = snake_case_ident(fn_descriptor.self_ty_name());
+
+    Ident::new(
+        &format!("{}_{}", self_ty_name, fn_descriptor.method_name),
```

Comment:
What do you think about using `some_struct__some_method` naming convention?

```suggestion
        &format!("{}__{}", self_ty_name, fn_descriptor.method_name),
```

Reply:
Or `some_struct___some_method`?

Reply:
Or `SomeStruct_some_method`?

Reply:
yes, I like the suggestion

Reply:
`SomeStruct__some_method` sounds even better